app/services/ipa.py

```diff
@@ -1,13 +1,117 @@
 from python_freeipa import Client
 from app.config import CONFIG, logger
+from typing import List
 import datetime
+import random
+import dns.resolver
 
+# --- Helper: DNS SRV Resolver ---
+def ipa_resolve_srv(service: str, protocol: str, domain: str) -> List[str]:
+    """
+    Resolves SRV records and returns A LIST of all valid hostnames,
+    sorted by Priority (asc) and then randomized by Weight/Shuffle.
 
-# --- Helper: IPA Client Wrapper ---
+    Args:
+        service: The service name (e.g., "_ldap")
+        protocol: The protocol (e.g., "_tcp")
+        domain: The domain name (e.g., "example.com")
+
+    Returns:
+        List of target hostnames strings (without trailing dots).
+        Returns empty list [] if no records found.
+    """
+    query_name = f"{service}.{protocol}.{domain}"
+    results = []
+
+    try:
+        answers = dns.resolver.resolve(query_name, 'SRV')
+
+        if not answers:
+            logger.info(f"No SRV records found for {query_name}")
+            return []
+
+        # Group records by priority
+        records_by_priority = {}
+        for rdata in answers:
+            prio = rdata.priority
+            if prio not in records_by_priority:
+                records_by_priority[prio] = []
+
+            # Store target (stripped of dot)
+            target = rdata.target.to_text().rstrip('.')
+            records_by_priority[prio].append(target)
+
+        # Process priorities in order (lowest number = highest priority)
+        sorted_priorities = sorted(records_by_priority.keys())
+
+        for prio in sorted_priorities:
+            candidates = records_by_priority[prio]
+
+            # Shuffle candidates of the same priority for basic load balancing
+            random.shuffle(candidates)
+
+            # Add them to the master list
+            results.extend(candidates)
+
+    except (dns.resolver.NoAnswer, dns.resolver.NXDOMAIN):
+        logger.info(f"No SRV records found for {query_name} or domain missing")
+    except Exception as e:
+        logger.info(f"An error occurred during SRV lookup: {e}")
+
+    return results
+
+
+# --- Helper: Get Authenticated Client (Retry Logic) ---
 def get_ipa_client():
-    c = Client(host=CONFIG["IPA_HOST"], verify_ssl=CONFIG["IPA_VERIFY_SSL"])
-    c.login(CONFIG["IPA_USER"], CONFIG["IPA_PASS"])
-    return c
+    """
+    Creates an authenticated Client by trying DNS candidates first,
+    then falling back to the static IPA_HOST config.
+    """
+    candidate_hosts = []
+
+    # 1. Try DNS Discovery (Dynamic)
+    # Note: Use CONFIG.get to safely access keys, assuming 'DOMAIN' is the key in your config
+    domain = CONFIG.get("DOMAIN") or CONFIG.get("domain")
```

Comment:
Is the lowercase check needed? Looks like this all gets converted to uppercase here: https://github.com/TheWizardOfOzolins/virt-joiner/blob/3ae4acd1abef1f78846e57fd6753cd53f72e96b7/app/config.py#L40-L41